test/edu/immutables/bst/TreeTest.java

```diff
@@ -0,0 +1,89 @@
+package edu.immutables.bst;
+
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.IntStream;
+
+import static org.junit.Assert.*;
+
+public class TreeTest {
+
+    private static class Console {
+        final List<String> buffer = new ArrayList<>();
+
+         <T> void out(final T object) {
+            buffer.add(object.toString());
+        }
+    }
+
+    @Test
+    public void add() throws Exception {
+        final Tree<Integer> tree = Tree.empty();
+        final Tree<Integer> newTree = tree.add(1);
+        assertEquals(1, newTree.size());
+        assertEquals(Collections.singletonList(1), newTree.toList());
+    }
+
+    @Test
+    public void remove() throws Exception {
+        final Tree<Integer> tree = Tree.of(1, 2, 3);
+        final Tree<Integer> newTree = tree.remove(1);
+        assertEquals(2, newTree.size());
+        assertEquals(Arrays.asList(2, 3), newTree.toList());
+    }
+
+    @Test
+    public void map() throws Exception {
+        final Tree<Integer> tree = Tree.of(1, 2, 3);
+        final Tree<String> newTree = tree.map(String::valueOf);
+        assertEquals(3, newTree.size());
+        assertEquals(Arrays.asList("1", "2", "3"), newTree.toList());
+    }
+
+    @Test
+    public void flatMap() throws Exception {
+        final Tree<Integer> tree = Tree.of(1, 2, 3);
+
+        final Tree<Integer> newTree = tree.flatMap(e ->
+                Tree.of(IntStream.range(1, e + 1).map(x -> e).boxed().toArray(Integer[]::new)));
+        assertEquals(6, newTree.size());
+        assertEquals(Arrays.asList(1, 2, 2, 3, 3, 3), newTree.toList());
+    }
+
+    @Test
+    public void foreach() throws Exception {
```

Comment:
why not just use a list here, you are even using a list behind the scenes !

Reply:
Sure will do it